pkg/config/release.go

```diff
@@ -163,6 +183,278 @@ func GetChangedTemplates(path, baseRev string) ([]ConfigMapSource, error) {
+	}
+	return nestedChainsSlice
+}
+
+func getReferencesInChains(steps []api.TestStep, fullRegistry registry) []string {
+	var usedReferences []string
+	for _, step := range steps {
+		if step.Chain != nil {
+			// this should never occur since we always provide unrolledChains; maybe throw an error (otherwise remove this if block)?
+			continue
+		}
+		if step.Reference != nil {
+			usedReferences = append(usedReferences, *step.Reference)
+		}
+	}
+	return usedReferences
+}
+
+func getStepsInWorkflowChanges(changes []RegistryStep) (refs, chains []string) {
+	for _, change := range changes {
+		if change.Workflow != nil {
+			steps := change.Workflow.Steps
+			combinedSteps := append(steps.Pre, steps.Test...)
+			combinedSteps = append(combinedSteps, steps.Post...)
+			for _, step := range combinedSteps {
+				if step.Chain != nil {
+					chains = append(chains, *step.Chain)
+				} else if step.Reference != nil {
+					refs = append(refs, *step.Reference)
+				}
+			}
+		}
+	}
+	return
+}
+
+// trimRegistryChanges removes changed registry components that are included as part of
+// other changed components, e.g. a changed reference that is part of a changed chain.
+// The trimming order is:
+// 1. Remove chains that exist in changed workflows from changed chains
+// 2. Remove any chains that are nested inside another changed chain (or chains inside changed workflows)
+// 3. Unroll all remaining chains and identify all used references
+// 4. Find all references used in workflows
+// 5. Remove all references that are included as part of a changed chain or workflow
+func trimRegistryChanges(changes []RegistryStep, fullRegistry registry) []RegistryStep {
+	wfRefs, wfChains := getStepsInWorkflowChanges(changes)
+	var chains []api.RegistryChain
+	for _, change := range changes {
+		if change.Chain != nil {
+			// only get chains not included in changed workflows
+			ignore := false
+			for _, chain := range wfChains {
+				if change.Chain.As == chain {
+					ignore = true
+				}
+			}
+			if !ignore {
+				chains = append(chains, *change.Chain)
+			}
+		}
+	}
+	// for getting nested chains, use both changed chains and chains from changed workflows
+	var resolvedWFChains []api.RegistryChain
+	for _, chain := range wfChains {
+		if resolvedChain, ok := fullRegistry.chains[chain]; !ok {
+			// throw error
+		} else {
+			// put into RegistryChain struct
+			registryChain := api.RegistryChain{As: chain, Steps: resolvedChain}
+			resolvedWFChains = append(resolvedWFChains, registryChain)
+		}
+	}
+	combinedChains := append(chains, resolvedWFChains...)
+	nestedChains := getNestedChains(combinedChains, fullRegistry)
+	var trimmedChanges []RegistryStep
+	// we still want a []api.TestStep of non-nested chains for later steps
+	chains = []api.RegistryChain{}
+	for _, change := range changes {
+		if change.Chain != nil {
+			ignore := false
+			for _, nested := range nestedChains {
+				if change.Chain.As == nested {
```

Comment:
All of these types of checks are cleaners with sets and `.Contains()` or `ContainsAny()`